flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableFsDataOutputStream.java

```diff
@@ -171,10 +189,20 @@ private static void safelyTruncateFile(
         }
 
         if (!truncated) {
-            // Truncate did not complete immediately, we must wait for
-            // the operation to complete and release the lease.
-            revokeLeaseByFileSystem(fileSystem, path);
+            // Truncate does not complete immediately if block recovery is needed,
+            // thus wait for block recovery
+            LOG.debug("Waiting for block recovery for file {}.", path);
+            boolean recovered = waitForBlockRecovery(fileSystem, path);
+            if (!recovered) {
+                throw new IOException(
+                        String.format(
+                                "Failed to recover blocks for file %s in %d ms.",
+                                path.toString(), BLOCK_RECOVERY_TIMEOUT_MS));
+            }
```

Comment:
Should the revokeLeaseByFileSystem() called after blockrecovery?

Reply:
IMHO, we've already triggered a lease recovery, so there's no need to call it again, which would interrupt the previous call.